File: lab1/src/T2.java
import java.util.Arrays;

public class T2 extends Thread {
    private int x2 = 0;

    public void run() {
        try {
            System.out.println("T2 is started");

            //Input MD
            for (int i = 0; i < Data.N; i++) {
                Arrays.fill(Data.MD[i], 1);
            }

            //Input d
            Data.d = 1;

            //Signal T1, T3, T4 about input MD, d
            Data.Sem2.release(3);

            //Wait for input MC, E in T1
            Data.Sem1.acquire(1);

            //Wait for input B, p in T3
            Data.Sem3.acquire(1);

            //Wait for input R, Z in T4
            Data.Sem4.acquire(1);

            //Calculation1 Sh = R * MCh
            Data.calculation1(Data.H, Data.H * 2);

            //Calculation2 x2 = BH * ZH
            for (int i = 0; i < Data.H; i++) {
                x2 += Data.B[i] * Data.Z[i];
            }

            //Calculation3 x = x + x2
            Data.x.addAndGet(x2);

            //Signal T1, T3, T4 about end of calculation x
            Data.Sem6.release(3);

            //Wait for end of calculation x in T1
            Data.Sem5.acquire(1);

            //Wait for end of calculation x in T3
            Data.Sem7.acquire(1);

            //Wait for end of calculation x in T4
            Data.Sem8.acquire(1);

            //Copy p2 = p
            int p2;
            synchronized (Data.Y) {
                p2 = Data.p;
            }

            //Copy x2 = x
            Data.Sem0.acquire();
            x2 = Data.x.get();
            Data.Sem0.release();

            //Copy d2 = d
            int d2;
            synchronized (Data.W) {
                d2 = Data.d;
            }

            //Calculation4 Ah = S * MDh * p2 + x2 * Eh * d2
            Data.calculation4(p2, x2, d2, 0, Data.H);

            //Signal T3 about end of calculation Ah
            Data.Sem10.release(1);

            System.out.println("T2 is finished");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: lab2/ConsoleApp1/T2.cs
namespace ConsoleApp1;

public class T2
{
    public static void Run(Data data)
    {
        Console.WriteLine("T2 is started");

        // Input B
        data.B = Enumerable.Repeat(1, data.N).ToArray();

        // Input MX
        data.MX = new int[data.N, data.N];
        for (var i = 0; i < data.N; i++)
        for (var j = 0; j < data.N; j++)
            data.MX[i, j] = 1;

        // Signal T1, T3, T4 about input B, MX
        data.Sem1.Release(3);

        // Wait for input d, Z, MM in T4
        data.Sem2.WaitOne();

        data.Mutex1.WaitOne();

        // Copy d2 = d
        var d2 = data.d;
        data.Mutex1.ReleaseMutex();

        // Calculation1 Ah = sort(d * Bh + Z * (MM * MXh))
        data.calculation1(d2, data.H, data.H * 2);

        // Signal T1 about end of calculation Ah
        data.Sem3.Release();

        // Calculation4 ai = min(Bh)
        var a2 = data.calculation4(data.H, data.H * 2);

        // Calculation5 a = min(a, ai)
        lock(data.aLock)
        {
            data.a = Math.Min(data.a, a2);
        }

        // Signal T1, T3, T4 about end of calculation a
        data.Event2.Set();

        // Wait for calculation a, A in T1
        data.Event1.WaitOne();

        // Wait for calculation a in T3
        data.Event3.WaitOne();

        // Wait for calculation a in T4
        data.Event4.WaitOne();

        data.Mutex2.WaitOne();
        // Copy a2 = a
        a2 = data.a;
        data.Mutex2.ReleaseMutex();

        // Calculation6 Xh = Ah * a
        data.calculation6(a2, data.H, data.H * 2);

        // Signal T4 about end of calculation Xh
        data.Barrier.SignalAndWait();

        Console.WriteLine("T2 is finished");
    }
}
